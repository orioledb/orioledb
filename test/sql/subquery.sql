CREATE SCHEMA subquery;
SET SESSION search_path = 'subquery';
CREATE EXTENSION orioledb;

CREATE TABLE o_test_subquery
(
  key bigint NOT NULL,
  val int,
  val2 int NOT NULL,
  PRIMARY KEY (key)
) USING orioledb;

CREATE INDEX o_test_subquery_idx1 ON o_test_subquery (val);
INSERT INTO o_test_subquery SELECT 1000 + i, 2000 + i, 3000 + i FROM generate_series(1, 500) AS i;
CREATE INDEX o_test_subquery_idx2 ON o_test_subquery (val2);

analyze o_test_subquery;

CREATE OR REPLACE FUNCTION smart_explain(sql TEXT) RETURNS SETOF TEXT AS $$
	DECLARE
		row RECORD;
		line text;
		indent integer;
		skip_indent integer;
		skip_start integer;
	BEGIN
		skip_indent := 0;
		skip_start := 0;
		FOR row IN EXECUTE sql LOOP
			line := row."QUERY PLAN";
			indent := length((regexp_match(line, '^ *'))[1]);
			IF line ~ '^ *->  Result' OR line ~ '^Result' THEN
				skip_indent := 6;
				skip_start := indent;
			ELSE
				IF indent >= skip_start THEN
					line := substr(line, skip_indent + 1);
				ELSE
					skip_indent := 0;
					skip_start := 0;
				END IF;
				RETURN NEXT line;
			END IF;
		END LOOP;
	END $$
LANGUAGE plpgsql;

-- index subscan; index only qual
SELECT smart_explain(
'EXPLAIN (COSTS off) WITH o_test_subquery_all AS (
    SELECT * FROM o_test_subquery
    WHERE val2 > 0 AND ABS(val2) IS NOT NULL
    ORDER BY val2
) SELECT COUNT(*) FROM o_test_subquery_all;');
-- returns 500
WITH o_test_subquery_all AS (
    SELECT * FROM o_test_subquery
    WHERE val2 > 0 AND ABS(val2) IS NOT NULL
    ORDER BY val2
) SELECT COUNT(*) FROM o_test_subquery_all;

-- index subscan; index only qual; query rows
SELECT smart_explain(
'EXPLAIN (COSTS off) WITH o_test_subquery_all AS (
    SELECT * FROM o_test_subquery
    WHERE val2 > 0 AND ABS(val2) IS NOT NULL
    ORDER BY val2
) SELECT * FROM o_test_subquery_all LIMIT 10;');
WITH o_test_subquery_all AS (
    SELECT * FROM o_test_subquery
    WHERE val2 > 0 AND ABS(val2) IS NOT NULL
    ORDER BY val2
) SELECT * FROM o_test_subquery_all LIMIT 10;

-- index subscan
SELECT smart_explain(
'EXPLAIN (COSTS off) WITH o_test_subquery_all AS (
    SELECT * FROM o_test_subquery
    WHERE val2 > 0 AND val > 0
    ORDER BY val2
) SELECT COUNT(*) FROM o_test_subquery_all;');
-- returns 500
WITH o_test_subquery_all AS (
    SELECT * FROM o_test_subquery
    WHERE val2 > 0 AND val > 0
    ORDER BY val2
) SELECT COUNT(*) FROM o_test_subquery_all;

-- index subscan; query rows
EXPLAIN (COSTS off) WITH o_test_subquery_all AS (
    SELECT * FROM o_test_subquery
    WHERE val2 > 0 AND val > 0
    ORDER BY val2
) SELECT * FROM o_test_subquery_all LIMIT 10;
WITH o_test_subquery_all AS (
    SELECT * FROM o_test_subquery
    WHERE val2 > 0 AND val > 0
    ORDER BY val2
) SELECT * FROM o_test_subquery_all LIMIT 10;

-- index only subscan; index only qual
EXPLAIN (COSTS off) WITH o_test_subquery_all AS (
    SELECT val2 FROM o_test_subquery
    WHERE val2 > 0 AND ABS(val2) IS NOT NULL
    ORDER BY val2
) SELECT COUNT(o_test_subquery_all.val2) FROM o_test_subquery_all;
-- returns 500
WITH o_test_subquery_all AS (
    SELECT val2 FROM o_test_subquery
    WHERE val2 > 0 AND ABS(val2) IS NOT NULL
    ORDER BY val2
) SELECT COUNT(o_test_subquery_all.val2) FROM o_test_subquery_all;

-- index only subscan; index only qual; query rows
EXPLAIN (COSTS off) WITH o_test_subquery_all AS (
    SELECT val2 FROM o_test_subquery
    WHERE val2 > 0 AND ABS(val2) IS NOT NULL
    ORDER BY val2
) SELECT o_test_subquery_all.val2 FROM o_test_subquery_all LIMIT 10;
WITH o_test_subquery_all AS (
    SELECT val2 FROM o_test_subquery
    WHERE val2 > 0 AND ABS(val2) IS NOT NULL
    ORDER BY val2
) SELECT o_test_subquery_all.val2 FROM o_test_subquery_all LIMIT 10;

-- index only subscan
EXPLAIN (COSTS off) WITH o_test_subquery_all AS (
    SELECT val2 FROM o_test_subquery
    WHERE val2 > 0 AND val > 4
    ORDER BY val2
) SELECT COUNT(o_test_subquery_all.val2) FROM o_test_subquery_all;

-- returns 500
WITH o_test_subquery_all AS (
    SELECT val2 FROM o_test_subquery
    WHERE val2 > 0 AND val > 4
    ORDER BY val2
) SELECT COUNT(o_test_subquery_all.val2) FROM o_test_subquery_all;

-- index only subscan; query rows
EXPLAIN (COSTS off) WITH o_test_subquery_all AS (
    SELECT val2 FROM o_test_subquery
    WHERE val2 > 0 AND val > 4
    ORDER BY val2
) SELECT o_test_subquery_all.val2 FROM o_test_subquery_all LIMIT 10;

WITH o_test_subquery_all AS (
    SELECT val2 FROM o_test_subquery
    WHERE val2 > 0 AND val > 4
    ORDER BY val2
) SELECT o_test_subquery_all.val2 FROM o_test_subquery_all LIMIT 10;

DROP EXTENSION orioledb CASCADE;
DROP SCHEMA subquery CASCADE;
RESET search_path;
