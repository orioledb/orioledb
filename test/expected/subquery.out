CREATE SCHEMA subquery;
SET SESSION search_path = 'subquery';
CREATE EXTENSION orioledb;
CREATE TABLE o_test_subquery
(
  key bigint NOT NULL,
  val int,
  val2 int NOT NULL,
  PRIMARY KEY (key)
) USING orioledb;
CREATE INDEX o_test_subquery_idx1 ON o_test_subquery (val);
INSERT INTO o_test_subquery SELECT 1000 + i, 2000 + i, 3000 + i FROM generate_series(1, 500) AS i;
CREATE INDEX o_test_subquery_idx2 ON o_test_subquery (val2);
analyze o_test_subquery;
CREATE OR REPLACE FUNCTION smart_explain(sql TEXT) RETURNS SETOF TEXT AS $$
	DECLARE
		row RECORD;
		line text;
		indent integer;
		skip_indent integer;
		skip_start integer;
	BEGIN
		skip_indent := 0;
		skip_start := 0;
		FOR row IN EXECUTE sql LOOP
			line := row."QUERY PLAN";
			indent := length((regexp_match(line, '^ *'))[1]);
			IF line ~ '^ *->  Result' OR line ~ '^Result' THEN
				skip_indent := 6;
				skip_start := indent;
			ELSE
				IF indent >= skip_start THEN
					line := substr(line, skip_indent + 1);
				ELSE
					skip_indent := 0;
					skip_start := 0;
				END IF;
				RETURN NEXT line;
			END IF;
		END LOOP;
	END $$
LANGUAGE plpgsql;
-- index subscan; index only qual
SELECT smart_explain(
'EXPLAIN (COSTS off) WITH o_test_subquery_all AS (
    SELECT * FROM o_test_subquery
    WHERE val2 > 0 AND ABS(val2) IS NOT NULL
    ORDER BY val2
) SELECT COUNT(*) FROM o_test_subquery_all;');
                            smart_explain                            
---------------------------------------------------------------------
 Aggregate
   ->  Index Only Scan using o_test_subquery_idx2 on o_test_subquery
         Index Cond: (val2 > 0)
         Filter: (abs(val2) IS NOT NULL)
(4 rows)

-- returns 500
WITH o_test_subquery_all AS (
    SELECT * FROM o_test_subquery
    WHERE val2 > 0 AND ABS(val2) IS NOT NULL
    ORDER BY val2
) SELECT COUNT(*) FROM o_test_subquery_all;
 count 
-------
   500
(1 row)

-- index subscan; index only qual; query rows
SELECT smart_explain(
'EXPLAIN (COSTS off) WITH o_test_subquery_all AS (
    SELECT * FROM o_test_subquery
    WHERE val2 > 0 AND ABS(val2) IS NOT NULL
    ORDER BY val2
) SELECT * FROM o_test_subquery_all LIMIT 10;');
                         smart_explain                          
----------------------------------------------------------------
 Limit
   ->  Index Scan using o_test_subquery_idx2 on o_test_subquery
         Index Cond: (val2 > 0)
         Filter: (abs(val2) IS NOT NULL)
(4 rows)

WITH o_test_subquery_all AS (
    SELECT * FROM o_test_subquery
    WHERE val2 > 0 AND ABS(val2) IS NOT NULL
    ORDER BY val2
) SELECT * FROM o_test_subquery_all LIMIT 10;
 key  | val  | val2 
------+------+------
 1001 | 2001 | 3001
 1002 | 2002 | 3002
 1003 | 2003 | 3003
 1004 | 2004 | 3004
 1005 | 2005 | 3005
 1006 | 2006 | 3006
 1007 | 2007 | 3007
 1008 | 2008 | 3008
 1009 | 2009 | 3009
 1010 | 2010 | 3010
(10 rows)

-- index subscan
SELECT smart_explain(
'EXPLAIN (COSTS off) WITH o_test_subquery_all AS (
    SELECT * FROM o_test_subquery
    WHERE val2 > 0 AND val > 0
    ORDER BY val2
) SELECT COUNT(*) FROM o_test_subquery_all;');
                         smart_explain                          
----------------------------------------------------------------
 Aggregate
   ->  Index Scan using o_test_subquery_idx2 on o_test_subquery
         Index Cond: (val2 > 0)
         Filter: (val > 0)
(4 rows)

-- returns 500
WITH o_test_subquery_all AS (
    SELECT * FROM o_test_subquery
    WHERE val2 > 0 AND val > 0
    ORDER BY val2
) SELECT COUNT(*) FROM o_test_subquery_all;
 count 
-------
   500
(1 row)

-- index subscan; query rows
EXPLAIN (COSTS off) WITH o_test_subquery_all AS (
    SELECT * FROM o_test_subquery
    WHERE val2 > 0 AND val > 0
    ORDER BY val2
) SELECT * FROM o_test_subquery_all LIMIT 10;
                           QUERY PLAN                           
----------------------------------------------------------------
 Limit
   ->  Index Scan using o_test_subquery_idx2 on o_test_subquery
         Index Cond: (val2 > 0)
         Filter: (val > 0)
(4 rows)

WITH o_test_subquery_all AS (
    SELECT * FROM o_test_subquery
    WHERE val2 > 0 AND val > 0
    ORDER BY val2
) SELECT * FROM o_test_subquery_all LIMIT 10;
 key  | val  | val2 
------+------+------
 1001 | 2001 | 3001
 1002 | 2002 | 3002
 1003 | 2003 | 3003
 1004 | 2004 | 3004
 1005 | 2005 | 3005
 1006 | 2006 | 3006
 1007 | 2007 | 3007
 1008 | 2008 | 3008
 1009 | 2009 | 3009
 1010 | 2010 | 3010
(10 rows)

-- index only subscan; index only qual
EXPLAIN (COSTS off) WITH o_test_subquery_all AS (
    SELECT val2 FROM o_test_subquery
    WHERE val2 > 0 AND ABS(val2) IS NOT NULL
    ORDER BY val2
) SELECT COUNT(o_test_subquery_all.val2) FROM o_test_subquery_all;
                             QUERY PLAN                              
---------------------------------------------------------------------
 Aggregate
   ->  Index Only Scan using o_test_subquery_idx2 on o_test_subquery
         Index Cond: (val2 > 0)
         Filter: (abs(val2) IS NOT NULL)
(4 rows)

-- returns 500
WITH o_test_subquery_all AS (
    SELECT val2 FROM o_test_subquery
    WHERE val2 > 0 AND ABS(val2) IS NOT NULL
    ORDER BY val2
) SELECT COUNT(o_test_subquery_all.val2) FROM o_test_subquery_all;
 count 
-------
   500
(1 row)

-- index only subscan; index only qual; query rows
EXPLAIN (COSTS off) WITH o_test_subquery_all AS (
    SELECT val2 FROM o_test_subquery
    WHERE val2 > 0 AND ABS(val2) IS NOT NULL
    ORDER BY val2
) SELECT o_test_subquery_all.val2 FROM o_test_subquery_all LIMIT 10;
                             QUERY PLAN                              
---------------------------------------------------------------------
 Limit
   ->  Index Only Scan using o_test_subquery_idx2 on o_test_subquery
         Index Cond: (val2 > 0)
         Filter: (abs(val2) IS NOT NULL)
(4 rows)

WITH o_test_subquery_all AS (
    SELECT val2 FROM o_test_subquery
    WHERE val2 > 0 AND ABS(val2) IS NOT NULL
    ORDER BY val2
) SELECT o_test_subquery_all.val2 FROM o_test_subquery_all LIMIT 10;
 val2 
------
 3001
 3002
 3003
 3004
 3005
 3006
 3007
 3008
 3009
 3010
(10 rows)

-- index only subscan
EXPLAIN (COSTS off) WITH o_test_subquery_all AS (
    SELECT val2 FROM o_test_subquery
    WHERE val2 > 0 AND val > 4
    ORDER BY val2
) SELECT COUNT(o_test_subquery_all.val2) FROM o_test_subquery_all;
                           QUERY PLAN                           
----------------------------------------------------------------
 Aggregate
   ->  Index Scan using o_test_subquery_idx2 on o_test_subquery
         Index Cond: (val2 > 0)
         Filter: (val > 4)
(4 rows)

-- returns 500
WITH o_test_subquery_all AS (
    SELECT val2 FROM o_test_subquery
    WHERE val2 > 0 AND val > 4
    ORDER BY val2
) SELECT COUNT(o_test_subquery_all.val2) FROM o_test_subquery_all;
 count 
-------
   500
(1 row)

-- index only subscan; query rows
EXPLAIN (COSTS off) WITH o_test_subquery_all AS (
    SELECT val2 FROM o_test_subquery
    WHERE val2 > 0 AND val > 4
    ORDER BY val2
) SELECT o_test_subquery_all.val2 FROM o_test_subquery_all LIMIT 10;
                           QUERY PLAN                           
----------------------------------------------------------------
 Limit
   ->  Index Scan using o_test_subquery_idx2 on o_test_subquery
         Index Cond: (val2 > 0)
         Filter: (val > 4)
(4 rows)

WITH o_test_subquery_all AS (
    SELECT val2 FROM o_test_subquery
    WHERE val2 > 0 AND val > 4
    ORDER BY val2
) SELECT o_test_subquery_all.val2 FROM o_test_subquery_all LIMIT 10;
 val2 
------
 3001
 3002
 3003
 3004
 3005
 3006
 3007
 3008
 3009
 3010
(10 rows)

DROP EXTENSION orioledb CASCADE;
NOTICE:  drop cascades to table o_test_subquery
DROP SCHEMA subquery CASCADE;
NOTICE:  drop cascades to function smart_explain(text)
RESET search_path;
